ding/reward_model/trex_reward_model.py

```diff
@@ -0,0 +1,342 @@
+        x = traj.permute(0, 3, 1, 2)  #get into NCHW format
+        #compute forward pass of reward network (we parallelize across frames so batch size is length of partial trajectory)
+        x = F.leaky_relu(self.conv1(x))
+        x = F.leaky_relu(self.conv2(x))
+        x = F.leaky_relu(self.conv3(x))
+        x = F.leaky_relu(self.conv4(x))
+        x = x.reshape(-1, 784)
+        x = F.leaky_relu(self.fc1(x))
+        r = self.fc2(x)
+        sum_rewards += torch.sum(r)
+        sum_abs_rewards += torch.sum(torch.abs(r))
+        return sum_rewards, sum_abs_rewards
+
+    def forward(self, traj_i, traj_j):
+        '''compute cumulative return for each trajectory and return logits'''
+        cum_r_i, abs_r_i = self.cum_return(traj_i)
+        cum_r_j, abs_r_j = self.cum_return(traj_j)
+        return torch.cat((cum_r_i.unsqueeze(0), cum_r_j.unsqueeze(0)), 0), abs_r_i + abs_r_j
+
+
+
+@REWARD_MODEL_REGISTRY.register('trex')
+class TrexRewardModel(BaseRewardModel):
+    """
+    Overview:
+        The Gail reward model class (https://arxiv.org/abs/1606.03476)
+    Interface:
+        ``estimate``, ``train``, ``load_expert_data``, ``collect_data``, ``clear_date``, \
+            ``__init__``, ``_train``,
+    """
+    config = dict(
+        type='trex',
+        learning_rate=1e-5,
+        # expert_data_path='expert_data.pkl'
+        update_per_collect=100,
+        batch_size=64,
+        # input_size=4,
+        target_new_data_count=64,
+        hidden_size=128,
+    )
+
+    def __init__(self, config: EasyDict, device: str, tb_logger: 'SummaryWriter') -> None:  # noqa
+        """
+        Overview:
+            Initialize ``self.`` See ``help(type(self))`` for accurate signature.
+        Arguments:
+            - cfg (:obj:`EasyDict`): Training config
+            - device (:obj:`str`): Device usage, i.e. "cpu" or "cuda"
+            - tb_logger (:obj:`SummaryWriter`): Logger, defaultly set as 'SummaryWriter' for model summary
+        """
+        super(TrexRewardModel, self).__init__()
+        self.cfg = config
+        assert device in ["cpu", "cuda"] or "cuda" in device
+        self.device = device
+        self.tb_logger = tb_logger
+        self.reward_model = TREX_Model()
+        self.reward_model.to(self.device)
+        self.pre_expert_data = []
+        self.train_data = []
+        self.expert_data_loader = None
+        self.opt = optim.Adam(self.reward_model.parameters(), config.reward_model.learning_rate)
+        self.train_iter = 0
+        self.learning_returns = []
+        self.learning_rewards = []
+        self.training_obs = []
+        self.training_labels = []
+        self.num_trajs = 0 # number of downsampled full trajectories
+        self.num_snippets = 6000 # number of short subtrajectories to sample
+        self.min_snippet_length = 50 # minimum number of short subtrajectories to sample
+        self.max_snippet_length = 100 # maximum number of short subtrajectories to sample
```

Comment:
It seems better to put these parameters to config because the length may be affect by cuda memory